### ISSUE TITLE
How to resolve these errors?

### ISSUE THREAD

**Issue by apptaro** [englercj/tsd-jsdoc#38]
Hi, @englercj I came here from: https://github.com/pixijs/pixi-ui/issues/16
Basically, I am trying to create .tsd for pixi-ui, but I see the errors below. How can I solve?

```
> npm install tsd-jsdoc
> npm install jsdoc@^3.4.0
> node_modules\.bin\jsdoc -t node_modules/tsd-jsdoc -r src
[TSD-JSDoc] Unable to resolve memberof for "PIXI.UI.Button", using memberof "PIXI.UI". No such name found.
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "String". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "String". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "String". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve memberof for "PIXI.UI.CheckBox", using memberof "PIXI.UI". No such name found.
[TSD-JSDoc] Unable to resolve type name "bool". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "String". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "String". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "String". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve memberof for "PIXI.UI.Container", using memberof "PIXI.UI". No such name found.
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Failed to resolve base type of "PIXI.UI.Container", no object found with name "PIXI.UI.UIBase".
[TSD-JSDoc] Unable to resolve memberof for "PIXI.UI.DynamicText", using memberof "PIXI.UI". No such name found.
[TSD-JSDoc] Unable to resolve type name "String". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "String". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "String". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "DynamicTextStyle". No type found with that name, defaulting to "any".
[TSD-JSDoc] Failed to resolve base type of "PIXI.UI.DynamicText", no object found with name "PIXI.UI.UIBase".
[TSD-JSDoc] Unable to resolve memberof for "PIXI.UI.InputBase", using memberof "PIXI.UI". No such name found.
[TSD-JSDoc] Failed to resolve base type of "PIXI.UI.InputBase", no object found with name "PIXI.UI.UIBase".
[TSD-JSDoc] Unable to resolve memberof for "PIXI.UI.ScrollBar", using memberof "PIXI.UI". No such name found.
[TSD-JSDoc] Unable to resolve type name "Object". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve memberof for "PIXI.UI.ScrollingContainer", using memberof "PIXI.UI". No such name found.
[TSD-JSDoc] Unable to resolve type name "Boolean". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Boolean". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Boolean". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "String". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Failed to resolve base type of "PIXI.UI.ScrollingContainer", no object found with name "PIXI.UI.UIBase".
[TSD-JSDoc] Unable to resolve memberof for "PIXI.UI.SliceSprite", using memberof "PIXI.UI". No such name found.
[TSD-JSDoc] Unable to resolve type name "PIXI.Texture". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Boolean". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Boolean". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Boolean". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve memberof for "PIXI.UI.Slider", using memberof "PIXI.UI". No such name found.
[TSD-JSDoc] Unable to resolve type name "Object". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "callback". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "callback". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve memberof for "PIXI.UI.SortableList", using memberof "PIXI.UI". No such name found.
[TSD-JSDoc] Unable to resolve type name "Boolean". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "PIXI.UI.Ease". No type found with that name, defaulting to "any".
[TSD-JSDoc] Failed to resolve base type of "PIXI.UI.SortableList", no object found with name "PIXI.UI.UIBase".
[TSD-JSDoc] Unable to resolve memberof for "PIXI.UI.Sprite", using memberof "PIXI.UI". No such name found.
[TSD-JSDoc] Unable to resolve type name "PIXI.Texture". No type found with that name, defaulting to "any".
[TSD-JSDoc] Failed to resolve base type of "PIXI.UI.Sprite", no object found with name "PIXI.UI.UIBase".
[TSD-JSDoc] Unable to resolve memberof for "PIXI.UI.Stage", using memberof "PIXI.UI". No such name found.
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve memberof for "PIXI.UI.Text", using memberof "PIXI.UI". No such name found.
[TSD-JSDoc] Unable to resolve type name "String". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "PIXI.TextStyle". No type found with that name, defaulting to "any".
[TSD-JSDoc] Failed to resolve base type of "PIXI.UI.Text", no object found with name "PIXI.UI.UIBase".
[TSD-JSDoc] Unable to resolve memberof for "PIXI.UI.TextInput", using memberof "PIXI.UI". No such name found.
[TSD-JSDoc] Unable to resolve type name "String". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Boolean". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "PIXI.TextStyle". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "String". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Array". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "String". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "String". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve memberof for "PIXI.UI.TilingSprite", using memberof "PIXI.UI". No such name found.
[TSD-JSDoc] Unable to resolve type name "PIXI.Texture". No type found with that name, defaulting to "any".
[TSD-JSDoc] Failed to resolve base type of "PIXI.UI.TilingSprite", no object found with name "PIXI.UI.UIBase".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "Number". No type found with that name, defaulting to "any".
[TSD-JSDoc] Failed to resolve base type of "UIBase", no object found with name "PIXI.UI.UIBase".
[TSD-JSDoc] Unable to resolve memberof for "PIXI.UI.UISettings", using memberof "PIXI.UI". No such name found.

```

**Comment by englercj** [comment]
> Unable to resolve type name ...

This means the jsdoc refers to a type that we didn't find anywhere. Generally this is because the jsdoc uses a type name that they didn't define. Or it is a [built-in type that we don't support yet](https://github.com/englercj/tsd-jsdoc/issues/16).

Specifically for `Number` I think what we look for is `number` so that might be the issue. Similarly for `Boolean` vs `boolean`, and `String` vs `string`, etc. You can see the built-in list here: https://github.com/RyanCavanaugh/dts-dom/blob/64dc5608ba9d497137614eebc90d4866d254e081/lib/index.ts#L561-L570

Other errors like this one:

> Unable to resolve memberof for "PIXI.UI.Button", using memberof "PIXI.UI". No such name found.

Are saying that the jsdoc for we looked for a container (namespace, class, etc) called `PIXI.UI` and found nothing. So the jsdoc has a memberof tag that points to something that was never defined anywhere.

**Comment by apptaro** [comment]
@englercj Thank you for your response.

I am new to JSDoc, so please help me here. For the Number/String/Boolean issue, should pixi-ui be fixed to use number/string/boolean for their docs?

**Comment by englercj** [comment]
Yes, either the jsdocs need to change or this lib needs to fix #16 or the other fix for this lib I mentioned.

**Comment by apptaro** [comment]
What about Object/Array? I get the error below even if I use lower-cased object/array:
```
[TSD-JSDoc] Unable to resolve type name "object". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "array". No type found with that name, defaulting to "any".

```

Also, pixi-ui references some types from pixijs, and gets the error below:
```
[TSD-JSDoc] Unable to resolve type name "PIXI.Texture". No type found with that name, defaulting to "any".
[TSD-JSDoc] Unable to resolve type name "PIXI.TextStyle". No type found with that name, defaulting to "any".

```
How do I let tsd-jsdoc know about those types?

**Comment by englercj** [comment]
> What about Object/Array? I get the error below even if I use lower-cased object/array:

Just like the error mentions, those types are not recognized, and so it falls back to `any`. Like mentioned in #16 built-in js types do not resolve currently. Either define the object explicitly with a typedef [like this](https://github.com/englercj/tsd-jsdoc/blob/84a6c4247cfeb31254d5cab99e96bac60c2c51e9/test/test.js#L20-L22) or change it to `any` or ignore the warning.

> Also, pixi-ui references some types from pixijs, and gets the error below:
> How do I let tsd-jsdoc know about those types?

This library isn't magic, it can only output information about what it knows about. It reads jsdoc and outputs equivalent typescript. If there is no jsdoc that tells the lib what `PIXI.Textrure` means, it can't output that type. So, tell it what the type is or use something else that it knows about.

**Comment by apptaro** [comment]
So issues addressed here is already on #16 so I'm closing this one.
